heron/schedulers/src/java/com/twitter/heron/scheduler/yarn/HeronMasterDriver.java

```diff
@@ -163,30 +165,86 @@ void scheduleHeronWorkers(PackingPlan topologyPacking) throws ContainerAllocatio
    * Container allocation is asynchronous. Requests all containers in the input set serially
    * to ensure allocated resources match the required resources.
    */
-  void scheduleHeronWorkers(Set<ContainerPlan> containers)
-      throws ContainerAllocationException {
+  void scheduleHeronWorkers(Set<ContainerPlan> containers) throws ContainerAllocationException {
     for (ContainerPlan containerPlan : containers) {
+      int id = containerPlan.getId();
       if (containerPlans.containsKey(containerPlan.getId())) {
-        throw new ContainerAllocationException("Received duplicate allocation request for "
-            + containerPlan.getId());
+        throw new ContainerAllocationException("Received duplicate allocation request for " + id);
       }
-      containerPlans.put(containerPlan.getId(), containerPlan);
-
       Resource reqResource = containerPlan.getRequiredResource();
-      int mem = getMemInMBForExecutor(reqResource);
-      int cores = getCpuForExecutor(reqResource);
-      launchContainerForExecutor(containerPlan.getId(), cores, mem);
+      containerPlans.put(id, containerPlan);
+      requestContainerForWorker(id, new HeronWorker(id, reqResource));
+    }
+  }
+
+  /**
+   * YARN allocates resources in fixed increments of memory and cpu. As a result, the actual
+   * resource allocation may be little more than what was requested. This method finds the biggest
+   * heron containerPlan that will fit the allocated YARN container. In some cases the YARN CPU
+   * scheduling may be disabled, resulting in default core allocation to each container. This
+   * method can ignore core fitting in such a case.
+   */
+  @VisibleForTesting
```

Comment:
No, the YARN scheduler continues to work with heterogeneous container sizes. YARN does not have a notion of request id in the allocated container. This method ensures that the allocated container is utilized properly by finding the biggest heron ContainerPlan that can fit in it.